### PR TITLE
Differentiate between epoch 409 and coherency 409 errors

### LINKED ISSUE #4762: Differentiate between epoch 409 error and other 409 errors.
## Describe the bug
Server throws 409 in a couple of cases, not all are epoch related. So make sure to only throw the epoch 409 error only when the server epoch does not match with the client epoch, so that it can be handled properly.

### PR DESCRIPTION
**Pull request by jatgarg**
Fixes: https://github.com/microsoft/FluidFramework/issues/4762

### CODE REVIEW

File: packages/drivers/odsp-driver/src/epochTracker.ts [hunk 1]
```diff
@@ -187,17 +183,37 @@ export class EpochTracker {
...
-            await this.persistedCache.removeEntries(this.fileEntry);
+            try {
+                // This will only throw if it is an epoch error.
+                this.checkForEpochErrorCore(epochFromResponse, error.errorMessage);
+            } catch (epochError) {
+                const err = {
+                    ...epochError,
+                    fromCache,
+                    clientEpoch: this.fluidEpoch,
+                    serverEpoch: epochFromResponse ?? undefined,
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/epochTracker.ts:194]: FYI - I've added serverEpcoh to every error in throwOdspNetworkError

File: packages/drivers/odsp-driver/src/epochTracker.ts [hunk 2]
```diff
@@ -187,17 +183,37 @@ export class EpochTracker {
...
+                this.logger.sendErrorEvent({ eventName: "EpochVersionMismatch" }, err);
+                assert(!!this.fileEntry, "File Entry should be set to clear the cached entries!!");
+                // If the epoch mismatches, then clear all entries for such file entry from cache.
+                await this.persistedCache.removeEntries(this.fileEntry);
+                throw epochError;
+            }
+            // If it was categorised as epoch error but the epoch returned in response matches with the client epoch
+            // then it was coherency 409, so rethrow it as throttling error so that it can retried. Default throttling
+            // time is 1s.
+            this.logger.sendTelemetryEvent({ eventName: "Coherency409" }, error);
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/epochTracker.ts:206]: I'd make it an error. Too many of these and the user experience is affected, so we want OCEs to see them quickly.

Comment (vladsud) [reply packages/drivers/odsp-driver/src/epochTracker.ts:206]: Also we might be wrong in our logic and server might indeed point to epoch failure :) And then knowing about it sooner would help us fix the problem sooner

File: packages/drivers/odsp-driver/src/epochTracker.ts (hunk 1 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/epochTracker.ts:194]: ok, then I will remove it from here.

File: packages/drivers/odsp-driver/src/epochTracker.ts [hunk 3]
```diff
@@ -187,17 +183,37 @@ export class EpochTracker {
...
-                serverEpoch: epochFromResponse ?? undefined,
-                fetchType,
-            };
-            this.logger.sendErrorEvent({ eventName: "EpochVersionMismatch" }, err);
-            assert(!!this.fileEntry, "File Entry should be set to clear the cached entries!!");
-            // If the epoch mismatches, then clear all entries for such file entry from cache.
-            await this.persistedCache.removeEntries(this.fileEntry);
+            try {
+                // This will only throw if it is an epoch error.
+                this.checkForEpochErrorCore(epochFromResponse, error.errorMessage);
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/epochTracker.ts:188]: It's a bit more refactoring (which is likely better to do in separate PR), but I think code would be easier to read if
1. checkForEpochErrorCore() would not throw but return an error object
2. Which suggest that most of the logic in throwOdspNetworkError should be moved into createOdspNetworkError and making throwOdspNetworkError  very shellow, and thus createOdspNetworkError() being used directly from checkForEpochErrorCore

What do you think?

**Review by vladsud — APPROVED** [review]

File: packages/drivers/odsp-driver/src/epochTracker.ts (hunk 3 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/epochTracker.ts:188]: Yes, I agree createOdspNetworkError  should do most of the work for creating the error and throwOdspNetworkError maybe just throw that error returned by create function. Currently both of them kind of add properties.

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/epochTracker.ts:188]: I will take that up in other PR.

**Pull request merged**